### PR TITLE
feat(prepro, ingest, deposition): Enforce author formatting in prepro, map authors accordingly in ingest and deposition

### LINKED ISSUE #2985: Author list formatting
ENA expects authors to be formatted as:
```
3.4.10.6  The RA Line
The RA (Reference Author) lines list the authors of the paper (or other 
work) cited. All of the authors are included, and are listed in the order 
given in the paper. The names are listed surname first followed by a blank
followed by initial(s) with stops. Occasionally the initials may not 
be known, in which case the surname alone will be listed. The author names 
are separated by commas and terminated by a semicolon; they are not split 
between lines. The RA line in the example is:
RA   Oxtoby E., Dunn M.A., Pancoro A., Hughes M.A.;    
As many RA lines as necessary are included for each reference.
```
see https://raw.githubusercontent.com/enasequence/read_docs/master/submit/fileprep/flatfile_user_manual.txt

As we need to submit AUTHORS in this format we need to make sure that users either submit AUTHORS in this format or in another format that we can easily convert into this format.

Lists of names separated by commas will not work as both Anna Maria Smith and Sascha Baron Cohen are valid names and there is no way to know if the middle name belongs to the last name or the first name. 

I would suggest we offer two author list formats:
1. standard EMBL format "surname firstname_initial1.firstname_initial2, surname firstname_initial;" if no firstname is known "surname"
2. custom format: "firstname, surname; firstname, surname;" if no firstname is known ", surname"

### PR DESCRIPTION
**Pull request by anna-parker**
resolves #2985, #1428

preview URL: https://format-authors.loculus.org/

##### Preprocessing changes

- From now on, we expect authors to be submitted in the new format: `Doe, John A.; Sanchez Roe, Jane Maria`, where last name(s) is mandatory, a comma is mandatory to separate first names/initials from last name.
- Only ASCII alphabetical characters A-Z are allowed in names
- Failure to do so results in the sequence becoming annotated with `error`, users will receive an error message with expected format
- Warn users if authors list might be in incorrect format.

#### Breaking: Loculus authors in released data/LAPIS

- The same format as used for submission will also be used for programmatic data output, instead of the current format: `FirstNames LastNames, FirstNames LastNames`

##### Ingest changes

- If authors list is entirely capitalized apply title case (leave common last names prefixes such as van etc. lower case)
- Add ingest test by modifying snakemake example files to also have edge cases.
- The tsv generated by the ncbi datasets cli from the raw jsonl does not correctly format authors- leading to the case where the string cannot be correctly parsed. To avoid this use the raw jsonl directly and write a python script to map json fields to a tsv This way we can format authors correctly, and have access to the raw data. 
- The following unused tsv metadata fields are dropped: 

```
ncbiHostBreed
ncbiHostCultivar
ncbiHostEcotype
ncbiHostIsolate
ncbiHostPangolin
ncbiIsVaccineStrain
ncbiMaturePeptideCount
ncbiMolType
ncbiVirusCommonName
ncbiVirusBreed
ncbiVirusCultivar
ncbiVirusEcotype
ncbiVirusIsolate
ncbi_virus (listed as "Virus Infraspecific Names Sex" in column_mapping)
ncbiVirusStrain
ncbiVirusPangolin
```

- The following currently always empty tsv fields are kept but left empty:

```
ncbiHostCommonName
ncbiPurposeOfSampling
```

To be accepted, ingest needs to resubmit all sequences

#### Problematic: Preprocessing version bump no longer possible

Old submissions won't be accepted by the new prepro pipeline, this is a problem for the current preprocessing versioning code which expects all released old sequences to be compatible with a new pipeline. There are a few options but none are obviously the best:
- A) Change author formatting inside the db sequence entries
- B) Make the prepro pipeline submission-time aware and be more lenient on old submissions before a time cutoff - but we'd still somehow need to ensure old entries have authors output in the new format, which is not trivial.  

### Summary
Preprocessing will only accept submissions where authors is formatted as `surname, first name; surname, first name;`. It will error if this is not the case.

- Ingest must now change how it formats the author string it receives from NIH to fit this format.
- Additionally, the ena-submission pod must map this new Loculus format to the ENA expected format.

### Screenshot 

example: old submitted author format that is no longer accepted
image

### Testing

Checkout https://github.com/pathoplexus/pathoplexus/commit/ce68b92e8b01055a36b5fa60893876078963bf98 to perform regression testing (ignore authors as should be different) by comparing this branch with Loculus main: 

```sh
micromamba activate pp-integrity
cd tests/regression-testing 
snakemake results/ebola-zaire.meta.main.format-authors.diff 
snakemake results/cchf.meta.main.format-authors.diff 
snakemake results/west-nile.meta.main.format-authors.diff 
```

There is no diff in the ebola-zaire, cchf and west nile metadata.

### PR Checklist

- [x] Include consistent white space
- [ ] Check if this causes a version bump in ingest (it should)
- [ ] Adjust documentation to include new author input format: https://github.com/pathoplexus/pathoplexus/pull/239
- [ ] Decide if we want to change the way we display authors on website
- [ ] Check that old sequences pass new prepro pipeline
- [x] Merge in https://github.com/loculus-project/loculus/pull/2990 so that ingested authors are in correct format

### CODE REVIEW

File: ena-submission/scripts/create_assembly.py [hunk 1]
```diff
@@ -140,9 +141,11 @@ def create_manifest_object(
         try:
             group_info = get_group_info(config, project_table_entry["group_id"])[0]["group"]
             address = group_info["address"]
-            address_string = (f'{address.get("line1", "")}, {address.get("line2", "")}, '
+            address_string = (
+                f'{address.get("line1", "")}, {address.get("line2", "")}, '
```
Comment (corneliusroemer) [inline ena-submission/scripts/create_assembly.py:145]: We probably actually want to add the institute name more prominently here - I don't think people actually care about the post code and address, at least that's what I saw when looking at what others submit.

Something like:
`Biozentrum, University of Basel` is more human friendly than `
Klingelbergstrasse 70, 4056 Basel, Switzerland` - what do you think? I know this is unrelated to this PR, just noticed - can make an issue.

File: ena-submission/scripts/ena_submission_helper.py [hunk 2]
```diff
@@ -129,6 +129,18 @@ def get_project_xml(project_set):
         "PROJECT": dataclass_to_xml(project_set, root_name="PROJECT_SET"),
     }
 
+def reformat_authors(authors: str) -> str:
```
Comment (corneliusroemer) [inline ena-submission/scripts/ena_submission_helper.py:132]:
```suggestion
def reformat_authors_from_loculus_to_embl_style(authors: str) -> str:
```

Maybe something like this is a bit more explicit - reformat could mean anything

File: ena-submission/scripts/ena_submission_helper.py [hunk 3]
```diff
@@ -129,6 +129,18 @@ def get_project_xml(project_set):
...
+    """This function reformats the Loculus authors string to the format expected by ENA"""
```
Comment (corneliusroemer) [inline ena-submission/scripts/ena_submission_helper.py:133]:
```suggestion
    """This function reformats the Loculus authors string to the format expected by ENA
    Loculus format: `John A. Doe; Jane B. C. Roe`
    EMBL expected: `Doe J.A., Roe J.B.C.`
    
    EMBL spec: "The names are listed surname first followed by a blank
      followed by initial(s) with stops.
      Occasionally the initials may not be known,
      in which case the surname alone will be listed.
      The author names are separated by commas
      and terminated by a semicolon; they are not split between lines."    
    See section "3.4.10.6: The RA Line" here: https://raw.githubusercontent.com/enasequence/read_docs/c4bd306c82710844128cdf43003a0167837dc442/submit/fileprep/flatfile_user_manual.txt"""
```

Adding some extra docs here to simplify checking this function and refactoring in future, also writing tests etc.

File: ena-submission/scripts/create_assembly.py (hunk 1 above)
Comment (anna-parker) [reply ena-submission/scripts/create_assembly.py:145]: I changed this in another PR: https://github.com/loculus-project/loculus/pull/2982/files

File: ena-submission/scripts/ena_submission_helper.py [hunk 4]
```diff
@@ -129,6 +129,18 @@ def get_project_xml(project_set):
...
+    authors_list = [author for author in authors.split(";") if author]
```
Comment (corneliusroemer) [inline ena-submission/scripts/ena_submission_helper.py:144]: What's the input format here? The one we show as authors on ppx?
`Nicole M Feriancek, Mallery I Breban, Ellie Bourgikos, Angela Bransfield, Philip M Armstrong, Chantal B.F. Vogels, Nathan D Grubaugh`

In that case splitting by `;` doesn't make sense? It must be something else. Maybe would be good to be more explicit what the expected input format is!

Comment (anna-parker) [reply ena-submission/scripts/ena_submission_helper.py:144]: ? This PR changes how we store authors in Loculus/PP so ena-submission will nolonger get authors in the format 
`Nicole M Feriancek, Mallery I Breban, Ellie Bourgikos, Angela Bransfield, Philip M Armstrong, Chantal B.F. Vogels, Nathan D Grubaugh` but in the format `Feriancek, Nicole M;` etc. If the ena-submission pod gets authors in an unexpected format it will throw an error and not submit the manifest

File: ena-submission/scripts/ena_submission_helper.py (hunk 3 above)
Comment (corneliusroemer) [reply ena-submission/scripts/ena_submission_helper.py]: Unclear about the Loculus input format used here.

Thinking about it, since we need to separate first names and last names, and last names can sometimes be multiple names that are whitespace separated, our format isn't unambiguous. Maybe we want to think about how we keep names unambiguous internally, something like:
`LASTNAME [OTHERLASTNAME...], FIRSTNAME [MOREFIRSTNAMES...]; ...` could work. Not a blocker for this PR, we can already improve based on our current system, just assuming single last names for now which should be correct in most cases (except for Spanish names)

File: ena-submission/scripts/ena_submission_helper.py (hunk 4 above)
Comment (anna-parker) [reply ena-submission/scripts/ena_submission_helper.py:144]: I'm confused why you are asking as you updated the function comment to include the expected format e.g. `John A. Doe; Jane B. C. Roe;`

File: ena-submission/scripts/ena_submission_helper.py (hunk 3 above)
Comment (corneliusroemer) [reply ena-submission/scripts/ena_submission_helper.py]: It seems that the above spec is exactly the input format you wrote the function for - but I'm not sure we're using that one at this point

File: ena-submission/scripts/ena_submission_helper.py [hunk 5]
```diff
@@ -129,6 +129,28 @@ def get_project_xml(project_set):
         "PROJECT": dataclass_to_xml(project_set, root_name="PROJECT_SET"),
     }
 
+def reformat_authors(authors: str) -> str:
+    """This function reformats the Loculus authors string to the format expected by ENA
+    Loculus format: `John A. Doe; Jane B. C. Roe`
```
Comment (anna-parker) [inline ena-submission/scripts/ena_submission_helper.py:134]:
```suggestion
    Loculus format: `John A. Doe; Jane B. C. Roe;`
```

File: ena-submission/scripts/ena_submission_helper.py (hunk 4 above)
Comment (corneliusroemer) [reply ena-submission/scripts/ena_submission_helper.py:144]: Sorry, I hadn't gotten to that part yet

File: ena-submission/scripts/ena_submission_helper.py [hunk 6]
```diff
@@ -129,6 +129,28 @@ def get_project_xml(project_set):
         "PROJECT": dataclass_to_xml(project_set, root_name="PROJECT_SET"),
     }
 
+def reformat_authors(authors: str) -> str:
+    """This function reformats the Loculus authors string to the format expected by ENA
+    Loculus format: `John A. Doe; Jane B. C. Roe;`
```
Comment (corneliusroemer) [inline ena-submission/scripts/ena_submission_helper.py:134]:
```suggestion
    Loculus format: `Doe, John A.; Roe, Jane B. C.;`
```

This is the new loculus format output by prepro - I hadn't realized that when I wrote the function comment.

File: ingest/scripts/prepare_metadata.py [hunk 7]
```diff
@@ -36,18 +36,18 @@ class Config:
 
 def split_authors(authors: str) -> str:
```
Comment (corneliusroemer) [inline ingest/scripts/prepare_metadata.py:37]:
```suggestion
def reformat_authors_from_genbank_to_loculus(authors: str) -> str:
```

The function comment probably needs to be rewritten as the operation is now much simpler: effectively we just need to replace every second comma with a semicolon `;`

File: ingest/scripts/prepare_metadata.py [hunk 8]
```diff
@@ -36,18 +36,18 @@ class Config:
...
 
     for i in range(0, len(single_split), 2):
         if i + 1 < len(single_split):
-            result.append(single_split[i + 1].strip() + " " + single_split[i].strip())
+            result.append(single_split[i].strip() + ", " + single_split[i + 1].strip())
         else:
             result.append(single_split[i].strip())
 
-    return ", ".join(result)
+    return "; ".join(result) + ";"
```
Comment (corneliusroemer) [inline ingest/scripts/prepare_metadata.py:50]: We can probably omit the trailing `;` - this isn't something that we need in Loculus - we can add it in the ENA submission code if it's needed there - it's also not in the docstring above, would have to be  `Xi, L.; Yu, X.;`

File: preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py [hunk 9]
```diff
@@ -45,6 +46,11 @@ def invalid_value_annotation(input_datum, output_field, value_type) -> Processin
     )
 
 
+def valid_authors(authors: str) -> bool:
+    pattern = r'^([a-zA-Z\s\.\-]+,[a-zA-Z\s\.\-]*;)+'
```
Comment (corneliusroemer) [inline preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py:50]:
```suggestion
    pattern = r'^([a-zA-Z\s\.\-]+,[a-zA-Z\s\.\-]*;)+'
```

This prohibits non-ascii characters - we may or may not want to accept them and then replace when submitting onwards to ENA (they definitely only allow ASCII). We can start restrictive though and become permissive later - we should just make this requirement clear in the docs and possibly make a discussion issue to discuss further.

Comment (corneliusroemer) [reply preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py:50]: We might also want to ensure that initials are terminated by a full stop, otherwise there'll be unnecessary inconsistency (we have that for West Nile right now:

```
Nicole M Feriancek, ..., Chantal B.F. Vogels,
```

Chantal has full stops after initials, the rest do not. We could require single first names to be followed by period (error), or rewrite to put the full stop there (warning), or inversely prohibit full stops altogether which would have the benefit of being more concise. 

The only issue is with single letter names which do occur if we were to accept Chinese characters (we probably shouldn't).

File: preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py [hunk 10]
```diff
@@ -400,6 +406,46 @@ def concatenate(
...
+                datum=authors,
+                warnings=warnings,
+                errors=errors,
+            )
+
+        error_message = (
+            f"The authors list '{authors}' are not in a recognized format. Please ensure that "
+            "authors are separated by semi-colons, and that each author’s last name and first "
+            "name are separated by a comma. "
+            "For example: 'lastname, firstname; lastname, firstname'."
```
Comment (corneliusroemer) [inline preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py:434]:
```suggestion
            "For example: 'lastname, firstname; lastname, firstname'."
```

We might want to add a more real example in particular one with multiple first names and/or initials

Comment (corneliusroemer) [reply preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py:434]: Also we might want to mention these edge cases in the docs and also here in the error message, otherwise it can be confusing

```
    "Xi,;",
```

I.e. last name only must still have comma, and first name only is not allowed.

File: ingest/scripts/prepare_metadata.py
Comment (corneliusroemer) [inline ingest/scripts/prepare_metadata.py:1]: This will bump versions in ingest - we could work around this with a more complicated prepro that takes into account whether something comes from ingest, but that might not be worth the hassle just to avoid +1 version.

File: ingest/scripts/prepare_metadata.py (hunk 8 above)
Comment (anna-parker) [reply ingest/scripts/prepare_metadata.py:50]: I think enforcing this will make it less likely that a user incorrectly just adds 2 names separated by a comma without realizing we have an expected author format, e.g. Anna Parker, Cornelius Roemer would be accepted by us and we would then think it is one name with surname=Anna Parker

File: preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py (hunk 10 above)
Comment (anna-parker) [reply preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py:434]: Ah yes this is lacking the end ; good catch!

File: ena-submission/scripts/ena_submission_helper.py [hunk 11]
```diff
@@ -129,6 +129,28 @@ def get_project_xml(project_set):
         "PROJECT": dataclass_to_xml(project_set, root_name="PROJECT_SET"),
     }
 
+def reformat_authors(authors: str) -> str:
+    """This function reformats the Loculus authors string to the format expected by ENA
+    Loculus format: `Doe, John A.; Roe, Jane B. C.;`
+    EMBL expected: `Doe J.A., Roe J.B.C.`
```
Comment (anna-parker) [inline ena-submission/scripts/ena_submission_helper.py:135]:
```suggestion
    EMBL expected: `Doe J.A., Roe J.B.C.;`
```
EMBL also seems to expect authors to end with a `;`

File: ingest/scripts/prepare_metadata.py [hunk 12]
```diff
@@ -36,18 +36,18 @@ class Config:
 
 def split_authors(authors: str) -> str:
     """Split authors by each second comma, then split by comma and reverse
-    So Xi,L.,Yu,X. becomes L. Xi, X. Yu
+    So Xi,L.,Yu,X. becomes  Xi, L.; Yu, X.
```
Comment (anna-parker) [inline ingest/scripts/prepare_metadata.py:39]:
```suggestion
    So Xi,L.,Yu,X. becomes  Xi, L.; Yu, X.;
```

File: ena-submission/scripts/ena_submission_helper.py (hunk 11 above)
Comment (corneliusroemer) [reply ena-submission/scripts/ena_submission_helper.py]: They expect yes, but that doesn't mean we need to expect that - their requirements are different as the flat file format is unstructured text as opposed to a TSV. We don't need to expect a trailing `;` - it makes code more complicated as we can't split/join by `;`, also makes things more verbose

File: preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py (hunk 9 above)
Comment (anna-parker) [reply preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py]: I have added an additional format_authors function to add in dots after initials - I will still add some tests now

File: preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py [hunk 13]
```diff
@@ -400,6 +423,49 @@ def concatenate(
...
+                warnings=warnings,
+                errors=errors,
+            )
+
+        error_message = (
+            f"The authors list '{authors}' is not in a recognized format. Please ensure that "
+            "authors are separated by semi-colons. Each author's name should be in the format "
+            "'last name, first name;'. Last name(s) is mandatory, a comma is mandatory to "
+            "separate first names/initials from last name. Only ASCII alphabetical characters A-Z"
+            "are allowed. For example: 'Smith, Anna; Perez, Tom J.; Xu, X.L.' "
```
Comment (anna-parker) [inline preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py:453]:
```suggestion
            "are allowed. For example: 'Smith, Anna; Perez, Tom J.; Xu, X.L.;' "
```

**Comment by anna-parker** [comment]
NIH data from NCBI virus is not consistent in their formatting of authors: 
```
13:18:17    DEBUG ( prepare_metadata.py:  83) - Config(compound_country_field='ncbiGeoLocation', fasta_id_field='genbankAccession', rename={'bioprojects': 'bioprojectAccession', 'country': 'geoLocCountry', 'division': 'geoLocAdmin1', 'genbankAccession': 'insdcAccessionFull', 'ncbiCollectionDate': 'sampleCollectionDate', 'ncbiHostCommonName': 'hostNameCommon', 'ncbiHostName': 'hostNameScientific', 'ncbiHostSex': 'hostGender', 'ncbiHostTaxId': 'hostTaxonId', 'ncbiIsLabHost': 'isLabHost', 'ncbiIsolateName': 'specimenCollectorSampleId', 'ncbiPurposeOfSampling': 'purposeOfSampling', 'ncbiSraAccessions': 'sraRunAccession', 'ncbiSubmitterAffiliation': 'authorAffiliations', 'ncbiSubmitterNames': 'authors'}, keep=['division', 'country', 'submissionId', 'insdcAccessionBase', 'insdcVersion', 'bioprojects', 'biosampleAccession', 'ncbiHostName', 'ncbiHostTaxId', 'ncbiIsLabHost', 'ncbiReleaseDate', 'ncbiUpdateDate', 'ncbiSourceDb', 'ncbiVirusName', 'ncbiVirusTaxId', 'sequence_md5', 'genbankAccession', 'jointAccession'], segmented=True)
13:18:17     INFO ( prepare_metadata.py:  85) - Reading metadata from results/filtered_metadata.tsv
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915541: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915542: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915543: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915544: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915545: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915546: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915547: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915548: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915549: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915550: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915551: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915552: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915553: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915554: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915555: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915556: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915557: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915558: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915559: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915560: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:17    ERROR ( prepare_metadata.py:  47) - Author list of MW915561: Shahid MF,Yaqub T,Ali M,Ul-Rahman A,Bente DA,Shahid,M.F.,Yaqub,T.,Ali,M.,Ul-Rahman,A.,Bente,D.A. has uneven number of first and last names, unable to format author names, returning empty author list
13:18:18     INFO ( prepare_metadata.py: 162) - Saved metadata for 3317 sequences
```

**Comment by anna-parker** [comment]
> MW915549

In the downloaded `ncbi_dataset/data/data_report.jsonl` the names are actually: 
"names":["Shahid MF","Yaqub T","Ali M","Ul-Rahman A","Bente DA","Shahid,M.F.","Yaqub,T.","Ali,M.","Ul-Rahman,A.","Bente,D.A."]}

File: ingest/scripts/prepare_metadata.py [hunk 14]
```diff
@@ -32,22 +34,59 @@ class Config:
...
+            msg = (
+                f"{insdc_accession_base}: Unexpected number of commas in author {author} "
+                f"in {authors}, not adding author to authors list"
+            )
+            logger.error(msg)
+            continue
+        formatted_authors.append(author_formatted)
+    return "; ".join(formatted_authors)
+
+def list_to_string(string_list: str) -> str:
```
Comment (anna-parker) [inline ingest/scripts/prepare_metadata.py:85]: move this to rename_fields.py

File: ingest/Snakefile [hunk 15]
```diff
@@ -72,30 +61,21 @@ rule fetch_ncbi_dataset_package:
...
+        config="results/config.yaml",
     output:
-        ncbi_dataset_tsv="results/metadata_post_extract.tsv",
+        ncbi_dataset_tsv="results/metadata_post_rename.tsv",
     shell:
         """
-        dataformat tsv virus-genome \
-            --package {input.dataset_package} \
-            > {output.ncbi_dataset_tsv}
+        unzip -o {input.dataset_package} -d results
```
Comment (anna-parker) [inline ingest/Snakefile:71]: It would probably be best to move the unzip into the python script to reduce points of failure

**Comment by anna-parker** [comment]
SQL code for retroactively updating the author format of previous versions:
```
WITH latest_authors AS (
    -- Find the latest version of each accession and extract its authors
    SELECT
        accession,
        metadata->'authors' AS latest_authors
    FROM entries
    WHERE (accession, version) IN (
        SELECT accession, MAX(version)
        FROM entries
        GROUP BY accession
    )
)
UPDATE entries
SET metadata = jsonb_set(metadata, '{authors}', latest_authors)
FROM latest_authors
WHERE entries.accession = latest_authors.accession
AND entries.version < (SELECT MAX(version) FROM entries e2 WHERE e2.accession = entries.accession);
```

File: ingest/Snakefile (hunk 15 above)
Comment (corneliusroemer) [reply ingest/Snakefile:71]: I think this comment was triggered by this bug: https://github.com/conda-forge/unzip-feedstock/issues/16

File: preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py [hunk 16]
```diff
@@ -45,6 +46,39 @@ def invalid_value_annotation(input_datum, output_field, value_type) -> Processin
...
+def warn_potentially_invalid_authors(authors: str) -> bool:
+    authors_split = re.split(r"[,\s]+", authors)
+    return bool(";" not in authors and len(authors_split) > 3)
+
+
+def format_authors(authors: str) -> bool:
+    authors_list = [author for author in authors.split(";") if author]
+    loculus_authors = []
+    for author in authors_list:
+        author_single_white_space = re.sub(r"\s\s+", " ", author)
```
Comment (corneliusroemer) [inline preprocessing/nextclade/src/loculus_preprocessing/processing_functions.py:63]:
```suggestion
        # Replace all multi-whitespace with single space
        author_single_white_space = re.sub(r"\s\s+", " ", author)
```

File: ena-submission/src/ena_deposition/config.py [hunk 17]
```diff
@@ -50,7 +50,7 @@ def secure_ena_connection(config: Config):
         config.test = True
         logging.info("Submitting to ENA dev environment")
         config.ena_submission_url = "https://wwwdev.ebi.ac.uk/ena/submit/drop-box/submit"
-        config.github_url = "https://raw.githubusercontent.com/pathoplexus/ena-submission/main/test/approved_ena_submission_list.json"
+        config.github_url = "https://raw.githubusercontent.com/pathoplexus/ena-submission/loculus_test/test/approved_ena_submission_list.json"
```
Comment (anna-parker) [inline ena-submission/src/ena_deposition/config.py:53]:
```suggestion
        config.github_url = "https://raw.githubusercontent.com/pathoplexus/ena-submission/main/test/approved_ena_submission_list.json"
```

File: backend/src/main/resources/db/migration/V1.4__fix_external_metadata_view.sql [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+            jsonb_build_object(
+                'metadata', 
+                (cpd.processed_data->'metadata') || all_external_metadata.external_metadata
+            )
+    end as joint_metadata
+from
+    (
+        -- Get only the preprocessed data for the current pipeline version
+        select * from sequence_entries_preprocessed_data 
+        where pipeline_version = (select version from current_processing_pipeline)
```
Comment (corneliusroemer) [inline backend/src/main/resources/db/migration/V1.4__fix_external_metadata_view.sql:24]: This is the key line here, need to filter _before_ join, as otherwise all processing versions end up in the view - Claude helped me figure this out!

Comment (anna-parker) [reply backend/src/main/resources/db/migration/V1.4__fix_external_metadata_view.sql:26]: Yes - this was my finding as well - I wrote it in the issue

File: ingest/config/defaults.yaml [hunk 19]
```diff
@@ -24,52 +24,44 @@ keep:
...
-  Virus Infraspecific Names Breed: ncbiVirusBreed
-  Virus Infraspecific Names Cultivar: ncbiVirusCultivar
-  Virus Infraspecific Names Ecotype: ncbiVirusEcotype
-  Virus Infraspecific Names Isolate: ncbiVirusIsolate
-  Virus Infraspecific Names Sex: ncbi_virus
-  Virus Infraspecific Names Strain: ncbiVirusStrain
-  Virus Name: ncbiVirusName
-  Virus Pangolin Classification: ncbiVirusPangolin
-  Virus Taxonomic ID: ncbiVirusTaxId
+simple_mappings:
```
[the remark this answers is not in the document]
Comment (anna-parker) [reply ingest/config/defaults.yaml:27]: Ah sorry should not have resolved: working on this in a follow-up PR: https://github.com/loculus-project/loculus/pull/3068

**Comment by anna-parker** [comment]
Lapis is still failing to start in the e2e tests - I looked into argocd and see it has started everywhere except for in the dummyOrganism, here I see silo-preprocessing has thrown an error:
```
[2024-10-25 10:35:58.073] [logger] [error] [api.cpp:289] Unknown metadata type: pango_lineage
[2024-10-25 10:35:58.073] [logger] [info] [api.cpp:328] Stopping SILO
std::exception
real	0m0.005s
user	0m0.000s
sys	0m0.004s
```

File: preprocessing/nextclade/src/loculus_preprocessing/backend.py [hunk 20]
```diff
@@ -77,7 +77,11 @@ def parse_ndjson(ndjson_data: str) -> Sequence[UnprocessedEntry]:
             continue
         # Loculus currently cannot handle non-breaking spaces.
         json_str_processed = json_str.replace("\N{NO-BREAK SPACE}", " ")
-        json_object = json.loads(json_str_processed)
+        try:
+            json_object = json.loads(json_str_processed)
+        except json.JSONDecodeError as e:
+            error_msg = f"Failed to parse JSON: {json_str_processed}"
+            raise Exception(error_msg) from e
```
Comment (corneliusroemer) [inline preprocessing/nextclade/src/loculus_preprocessing/backend.py:84]: We might actually want to catch this and not crash prepro as a result, just log an error - it's a fault of the backend, not a fault in the prepro pipeline.

**Review by corneliusroemer — APPROVED** [review]
